### PR TITLE
UI Editor

### LINKED ISSUE #5: UI Editor
Provide a UI Editor that is bundled in the final release

### PR DESCRIPTION
**Pull request by iantrich**
- [x] Figure out how best to approach replicating `fireEvent` 
- [ ] Figure out how to bundle multiple files

Fixes #5

### CODE REVIEW

**Comment by ljmerza** [comment]
I assume the commented out import is what's wrong? We could change the rollup.config `output.file` to `output.dir` but then you'd loose the file names. I can get build to to work with `output: {dir: './', format: 'es'}` but then the output files are `chunk[hash].js`.  Doesn't look like rollup is adhering to webpack magic comments. I bet not code splitting would work so there's only one output file

**Comment by iantrich** [comment]
Correct on everything. Code splitting would probably be best

**Comment by ljmerza** [comment]
Ok looks like we'd have to use the rollup api directly for this one. I got it to create two files - each have litElement bundled

```
const rollup = require('rollup');
const typescript = require('rollup-plugin-typescript2');
const resolve = require('rollup-plugin-node-resolve');


const cardOptions = {
  input: {
    input: 'src/index.ts',
    plugins: [resolve(), typescript()],
    inlineDynamicImports: true
  },
  output: {
    file: './boilerplate-card.js',
    format: 'es',
  },
};


const editorOptions = {
  input: {
    input: 'src/editor.ts',
    plugins: [resolve(), typescript()],
    inlineDynamicImports: true
  },
  output: {
    file: './boilerplate-card-editor.js',
    format: 'es',
  },
};

async function build() {
  const bundle = await rollup.rollup(cardOptions.input);
  await bundle.write(cardOptions.output);

  const bundleEditor = await rollup.rollup(editorOptions.input);
  await bundleEditor.write(editorOptions.output)
}

build();
```

then you'd just run `node rollup.config.js`

**Comment by ljmerza** [comment]
This way seems a little closer

```
import resolve from 'rollup-plugin-node-resolve';
import typescript from 'rollup-plugin-typescript2';

export default {
  input: ['src/editor.ts', 'src/index.ts'],
  output: {
    dir: './dist',
    format: 'es',
  },
  plugins: [resolve(), typescript()],
}
```

then `npm run build`. It makes three files, editor.js, index.js, and a chunk file (I assume lit element). The code isn't duplicated like the api config above.

File: src/editor.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  customElement,
+  property,
+  css,
+  CSSResult,
+} from 'lit-element';
+import '@polymer/paper-input/paper-input';
+import '@polymer/paper-toggle-button/paper-toggle-button';
+import { superstruct } from 'superstruct';
+
+import { BoilerplateConfig } from './index';
```
Comment (ljmerza) [inline src/editor.ts]: this is a circular dependency. maybe we can move BoilerplateConfig to its own file?

File: src/boilerplate-card-editor.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  private _valueChanged(ev: Event): void {
+    if (!this._config || !this.hass) {
+      return;
+    }
+    const target = ev.target! as any;
+
+    if (
+      this[`_${target.configValue}`] === target.value ||
```
Comment (ljmerza) [inline src/boilerplate-card-editor.ts]: I did this as well on the light-entity card at first but say you change the temperature to 100 then decide to change the color to . The color temperature isn't valid anymore but it's still at 100 so if I wanted to go back to the previous color temperature I want to send the same value it was previously. Something to think about at least.

**Pull request closed without merging**